hash/hash.go

```diff
@@ -0,0 +1,127 @@
+package hash
+
+import (
+	"bytes"
+	"crypto/subtle"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/NebulousLabs/fastrand"
+	"golang.org/x/crypto/argon2"
+)
+
+var (
+	ErrInvalidHash               = errors.New("the encoded hash is not in the correct format")
+	ErrIncompatibleVersion       = errors.New("incompatible version of argon2")
+	ErrMismatchedHashAndPassword = errors.New("passwords do not match")
+
+	// config is the configuration of the argon2id hasher.
+	// For the moment we'll use the setup we had with Kratos. Later on we can
+	// move this to a configuration file, if needed.
+	config = argon2Config{
+		SaltLength:  16,
+		Iterations:  2,
+		Memory:      131072,
+		Parallelism: 1,
+		KeyLength:   16,
+	}
+)
+
+type argon2Config struct {
+	SaltLength  uint32
+	Iterations  uint32
+	Memory      uint32
+	Parallelism uint8
+	KeyLength   uint32
+}
+
+// Generate hashes the given password with the default configuration.
+func Generate(password []byte) ([]byte, error) {
+	salt := fastrand.Bytes(int(config.SaltLength))
+	hash := argon2.IDKey(password, salt, config.Iterations, config.Memory, config.Parallelism, config.KeyLength)
```

Comment:
I think it would be very confusing for everyone reading this who's not really into the matter. Password hashing is a familiar concept to most people and throwing the word "key" at them would only create confusion. I, myself, don't know how argon2 works under the hood and I haven't looked into how this "key" is generated and used.